### PR TITLE
[14.0][MIG] sale_procurement_amendment

### PR DESCRIPTION
**Pull request by rousseldenis**
Depends on:

- [x] #1572

### CODE REVIEW

File: sale_procurement_amendment/README.rst [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+In Odoo, when you confirm a Sale Order, you can increase the quantities
+in sale lines. This is reflected in Pickings.
+But, if you decrease the quantities, you have to do this manually even if
+nothing has been done in any picking.
+
+This allows to reflect confirmed sale lines quantities
+amendments to procurements (and pickings) if nothing has been already begun
```
Comment (gurneyalex) [inline sale_procurement_amendment/README.rst:34]:
```suggestion
amendments to procurements (and pickings) if nothing was already started
```

File: sale_procurement_amendment/models/sale_order_line.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+
+class SaleOrderLine(models.Model):
+
+    _inherit = "sale.order.line"
+
+    to_be_procured = fields.Boolean(
+        compute="_compute_to_be_procured",
+        help="Technical field that check if the whole order line quantities"
+        "have the corresponding stock moves quantities in order to fullfill"
```
Comment (gurneyalex) [inline sale_procurement_amendment/models/sale_order_line.py:15]:
```suggestion
        help="Technical field that check if the whole order line quantities "
        "have the corresponding stock moves quantities in order to fullfill "
```

File: sale_procurement_amendment/models/sale_order_line.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        "move_ids",
+        "move_ids.state",
+        "move_ids.product_qty",
+        "product_uom_qty",
+    )
+    def _compute_to_be_procured(self):
+        """
+        This compare the current procurement quantities (see Odoo
+        _action_procurement_create function)
+        :return:
```
Comment (gurneyalex) [inline sale_procurement_amendment/models/sale_order_line.py:73]:
```suggestion
```

File: sale_procurement_amendment/models/sale_order_line.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    @api.depends(
+        "order_id.picking_ids.can_be_amended",
+        "chained_move_ids.picking_id.can_be_amended",
+    )
+    def _compute_pickings_in_progress(self):
+        """
+        Compute the picking in progress. That depends on picking
+        'can_be_amended' state. If one is not in the 'can_be_amended' state,
+        we are in the 'pickings_in_progress' situation.
+        :return:
```
Comment (gurneyalex) [inline sale_procurement_amendment/models/sale_order_line.py:52]:
```suggestion
```

File: sale_procurement_amendment/models/sale_order_line.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        compute="_compute_can_amend_and_reprocure",
+    )
+
+    @api.depends("pickings_in_progress", "product_id.type")
+    def _compute_can_amend_and_reprocure(self):
+        """
+            Filter here the lines that can be reprocured
+
+        :return: [description]
+        :rtype: [type]
```
Comment (gurneyalex) [inline sale_procurement_amendment/models/sale_order_line.py:34]:
```suggestion
```

File: sale_procurement_amendment/models/stock_move.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    )
+
+    @api.depends("state", "move_line_ids.qty_done")
+    def _compute_can_be_amended(self):
+        """
+        Look into operations in progress
+        As there is one operation in progress, move cannot be modified
+        as it will recompute all operations. User should start from
+        the beginning, that's unwanted
+        :return:
```
Comment (gurneyalex) [inline sale_procurement_amendment/models/stock_move.py:23]:
```suggestion
```

File: sale_procurement_amendment/models/stock_picking.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+    can_be_amended = fields.Boolean(
+        compute="_compute_can_be_amended",
+    )
+
+    @api.depends("move_lines.can_be_amended")
+    def _compute_can_be_amended(self):
+        """
+        Look if all moves in the picking can be modified
+        :return:
```
Comment (gurneyalex) [inline sale_procurement_amendment/models/stock_picking.py:19]:
```suggestion
```

File: sale_procurement_amendment/models/stock_picking.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+class StockPicking(models.Model):
+
+    _inherit = "stock.picking"
+
+    can_be_amended = fields.Boolean(
+        compute="_compute_can_be_amended",
+    )
+
+    @api.depends("move_lines.can_be_amended")
```
Comment (gurneyalex) [inline sale_procurement_amendment/models/stock_picking.py:15]:
```suggestion
    @api.depends("move_lines.can_be_amended", "move_lines.state")
```

**Review by gurneyalex — APPROVED** [review]
some cleanup proposed.

**Comment by rousseldenis** [comment]
> /ocabot merge nobump

@dreispt Rebase has raised a problem in tests. Need to fix before merge

**Comment by rousseldenis** [comment]
@dreispt Ready

**Pull request merged**